airbyte-ci/connectors/pipelines/pipelines/dagger/actions/gradle.py

```diff
@@ -0,0 +1,216 @@
+                )
+            )
+            # Set RUN_IN_AIRBYTE_CI to tell gradle how to configure its build cache.
+            # This is consumed by settings.gradle in the repo root.
+            .with_env_variable("RUN_IN_AIRBYTE_CI", "1")
+            # Disable the Ryuk container because it needs privileged docker access which it can't have.
+            .with_env_variable("TESTCONTAINERS_RYUK_DISABLED", "true")
+            .with_workdir(self.workdir)
+            .with_mounted_directory(self.workdir, self.gradle_project_dir)
+        )
+
+    @staticmethod
+    def set_s3_build_cache_credentials(
+        s3_build_cache_access_key_id: Optional[dagger.Secret], s3_build_cache_secret_key: Optional[dagger.Secret]
+    ) -> Callable:
+        """Closure to set the S3 build cache credentials as environment variables on a gradle container.
+
+        Args:
+            s3_build_cache_access_key_id (Optional[dagger.Secret]): The S3 build cache access key ID as a dagger secret.
+            s3_build_cache_secret_key (Optional[dagger.Secret]): The S3 build cache secret key as a dagger secret.
+
+        Returns:
+            Callable: A function which takes a gradle container and returns a gradle container with the S3 build cache credentials set as environment variables.
+        """
+
+        def wrapper(gradle_container: dagger.Container) -> dagger.Container:
+            if s3_build_cache_access_key_id:
+                gradle_container = gradle_container.with_secret_variable("S3_BUILD_CACHE_ACCESS_KEY_ID", s3_build_cache_access_key_id)
+            if s3_build_cache_secret_key:
+                gradle_container = gradle_container.with_secret_variable("S3_BUILD_CACHE_SECRET_KEY", s3_build_cache_secret_key)
+            return gradle_container
+
+        return wrapper
+
+    @staticmethod
+    def seed_gradle_cache_from_volume(gradle_dep_cache_path: str, gradle_home_path: str, local_maven_repository_path: str) -> Callable:
+        """Closure to seed the gradle home cache from the cache volume.
+        Args:
+            gradle_dep_cache_path (str): The path to the gradle dependency cache volume mount.
+            gradle_home_path (str): The path to the gradle home directory.
+            local_maven_repository_path (str): The path to the local maven repository.
+
+        Returns:
+            Callable: A function which takes a gradle container and returns a gradle container with the gradle cache seeded from the volume.
+        """
+
+        def wrapper(gradle_container: dagger.Container) -> dagger.Container:
+            return gradle_container.with_exec(
+                sh_dash_c(
+                    [
+                        # Ensure that the .m2 directory exists.
+                        f"mkdir -p {local_maven_repository_path}",
+                        # Copy the gradle cache from the volume to the gradle home.
+                        f"(rsync -a --stats --mkpath {gradle_dep_cache_path}/ {gradle_home_path} || true)",
```

Comment:
You can't break up the `with_exec` block formerly defined in `with_whole_git_repo`; it won't work, because dagger's layer caching will get in the way. It's so painful to get right, trust me.

Reply:
Oh, right.
So I should always chain rsync + gradle exec when running a gradle task right?
```
sh_dash_c(
                    [
                        # Ensure that the .m2 directory exists.
                        f"mkdir -p {self.LOCAL_MAVEN_REPOSITORY_PATH}",
                        # Load from the cache volume.
                        f"(rsync -a --stats --mkpath {self.GRADLE_DEP_CACHE_PATH}/ {self.GRADLE_HOME_PATH} || true)",
                        # Resolve all dependencies and write their checksums to './gradle/verification-metadata.dryrun.xml'.
                        self._get_gradle_command("help", *warm_dependency_cache_args),
                        # Build the CDK and publish it to the local maven repository.
                        self._get_gradle_command(":airbyte-cdk:java:airbyte-cdk:publishSnapshotIfNeeded"),
                        # Store to the cache volume.
                        f"(rsync -a --stats {self.GRADLE_HOME_PATH}/ {self.GRADLE_DEP_CACHE_PATH} || true)",
                    ]
                )
```

Reply:
More accurately: you want exactly that block above to run before running any other gradle task. Those subsequent tasks don't need to rsync because the S3 build cache will take over.